File: backend/src/analyzers/quickest_mate.py
"""
Unified analyzer for finding the quickest checkmate win (excluding Qxf7# scholar's mates).
Uses the unified move-by-move approach for efficiency.
"""
import chess
import re
from typing import List, Dict, Any, Optional
from ..unified_analyzer import UnifiedAnalyzerBase, MoveContext
from ..game_data import GameData


class UnifiedQuickestMateAnalyzer(UnifiedAnalyzerBase):
    """
    Unified version of QuickestMateAnalyzer.
    Finds the fastest checkmate win (excluding scholar's mate pattern).
    Tracks the single fastest mate across all games.
    Uses unified move-by-move processing for better performance.
    """
    
    def __init__(self, username: str):
        """Initialize with the username to filter for."""
        super().__init__(username)
        self.quickest_mate = float('inf')  # Track fastest mate (in half-moves)
        self.quickest_game_ref = None  # Lightweight reference to best game
    
    def start_game(self, game_data: GameData, user_is_white: bool, user_is_black: bool):
        """
        Initialize analyzer state for a new game.
        
        Args:
            game_data: The game data
            user_is_white: Whether user is playing white
            user_is_black: Whether user is playing black
        """
        super().start_game(game_data, user_is_white, user_is_black)
        
        # Track state for this game
        self.game_ended_in_mate = False
        self.game_move_count = 0
    
    def process_move(self, context: MoveContext):
        """Process a single move. Track move count."""
        self.game_move_count += 1
    
    def finish_game(self) -> List[Dict[str, Any]]:
        """
        Finalize analysis for the game and return findings.
        Called after all moves have been processed.
        
        Checks if the game ended in checkmate and if it's faster than current best.
        
        Returns:
            List of findings for this game (empty - we track across all games)
        """
        # Check if user won
        result = self.game_data.metadata.result
        user_won = (result == "1-0" and self.user_is_white) or (result == "0-1" and self.user_is_black)
        if not user_won:
            return []
        
        # Check if game ended in checkmate
        moves = self.game_data.moves
        if not moves or '#' not in moves[-1]:
            return []
        
        last_move = moves[-1]
        
        # Exclude Qxf7# or Qf7# (scholar's mate pattern)
        if last_move.startswith('Q') and 'f7' in last_move and '#' in last_move:
            return []
        # Also exclude Qxf2# for black's version
        if last_move.startswith('Q') and 'f2' in last_move and '#' in last_move:
            return []
        
        move_count = len(moves)
        
        # Check if this is faster than current best
        if move_count < self.quickest_mate:
            self.quickest_mate = move_count
            self.quickest_game_ref = {
                "game_data": self.game_data,
                "move_count": move_count,
                "last_move": last_move,
                "user_is_white": self.user_is_white
            }
        
        return []  # Return empty - we track across all games
    
    def get_game_points(self, config: dict) -> int:
        """Return points for current game based on existing findings."""
        # Quickest mate doesn't contribute to best game scoring
        # (it's a stat, not a feature to score)
        return 0
    
    def _extract_fen(self, pgn: str) -> Optional[str]:
        """Extract FEN from PGN headers (CurrentPosition or FEN)."""
        # Look for FEN in headers
        match = re.search(r'\[FEN\s+"([^"]+)"\]', pgn)
        if match:
            return match.group(1)
        # Chess.com includes CurrentPosition for final position
        match = re.search(r'\[CurrentPosition\s+"([^"]+)"\]', pgn)
        if match:
            return match.group(1)
        return None
    
    def _extract_elo(self, pgn: str, elo_header: str) -> Optional[int]:
        """Extract ELO rating from PGN header."""
        match = re.search(rf'\[{elo_header}\s+"(\d+)"\]', pgn)
        if match:
            try:
                return int(match.group(1))
            except ValueError:
                return None
        return None
    
    def get_final_results(self) -> List[Dict[str, Any]]:
        """
        Get final results after processing all games.
        Returns the single fastest checkmate found.
        NOW extracts FEN and ELO only for the selected result.
        
        Returns:
            List containing the fastest checkmate finding (or empty if none found)
        """
        if not self.quickest_game_ref:
            return []
        
        ref = self.quickest_game_ref
        
        # Extract ELO (deferred until here)
        white_elo = self._extract_elo(ref["game_data"].pgn, "WhiteElo")
        black_elo = self._extract_elo(ref["game_data"].pgn, "BlackElo")
        
        # Determine winner and loser
        winner_is_white = ref["user_is_white"]
        winner = ref["game_data"].metadata.white if winner_is_white else ref["game_data"].metadata.black
        loser = ref["game_data"].metadata.black if winner_is_white else ref["game_data"].metadata.white
        
        # Use half-move count - convert to 0-indexed (last move is at len(moves) - 1)
        # move_count is the total number of half-moves, so last move index is move_count - 1
        half_move_number = ref["move_count"] - 1 if ref["move_count"] > 0 else 0
        
        # Build replay data structure for frontend (uses config file - 0/0 = full game)
        from ..replay_helper import build_replay_data
        replay_data = build_replay_data(
            game_data=ref["game_data"],
            key_half_move=0,  # Start from move 0 like favorite_gambit
            feature_name="quickest_mate"  # Loads settings from replay_config.json (0/0 = full game)
        )
        
        # Build full finding with extracted data
        finding = {
            "feature_name": "quickest_mate",
            "display_name": "Quickest Mate",
            "game_metadata": {
                "white": ref["game_data"].metadata.white,
                "black": ref["game_data"].metadata.black,
                "link": ref["game_data"].metadata.link,
                "white_elo": white_elo,  # Extracted here, not during processing!
                "black_elo": black_elo,  # Extracted here, not during processing!
                "user_color": "white" if ref["user_is_white"] else "black",
                "pgn": ref["game_data"].pgn,
                # New simplified format for frontend navigation
                "all_moves": replay_data["all_moves"],
                "key_position_index": replay_data["key_position_index"],
                "fen": replay_data["fen"]
            },
            "position_link": f"{ref['game_data'].metadata.link}?move={half_move_number}" if ref["game_data"].metadata.link else None,
            "result_data": {
                "move_count": {
                    "value": (ref["move_count"] + 1) // 2,  # Display full move number, not half-moves
                    "label": "Moves to Mate"
                }
            }
        }
        
        return [finding]

File: backend-rust/crates/chess-analyzers/src/analyzers/quickest_mate.rs
use crate::analyzer_trait::{GameAnalyzer, MoveContext};
use chess_core::game_data::GameData;
use shakmaty::Position;

/// Tracks the fewest-move checkmate across all games. Excludes scholar's mate
/// (Qxf7# or Qf7#). Reports only the single quickest.
pub struct QuickestMateAnalyzer {
    username: String,
    matched_links: Vec<String>,
    current_link: Option<String>,
    user_is_white: bool,
    // Track the quickest mate found so far
    quickest_moves: Option<usize>,
    quickest_link: Option<String>,
    // Current game state
    current_mate_move: Option<usize>,
    is_scholars_mate: bool,
    total_moves: usize,
}

impl QuickestMateAnalyzer {
    pub fn new(username: &str) -> Self {
        Self {
            username: username.to_lowercase(),
            matched_links: Vec::new(),
            current_link: None,
            user_is_white: true,
            quickest_moves: None,
            quickest_link: None,
            current_mate_move: None,
            is_scholars_mate: false,
            total_moves: 0,
        }
    }
}

impl GameAnalyzer for QuickestMateAnalyzer {
    fn name(&self) -> &'static str {
        "quickest_mate"
    }

    fn start_game(&mut self, game_data: &GameData, user_is_white: bool) {
        self.current_link = game_data.metadata.link.clone();
        self.user_is_white = user_is_white;
        self.current_mate_move = None;
        self.is_scholars_mate = false;
        self.total_moves = 0;

        // Check for scholar's mate pattern in SAN moves
        // Scholar's mate typically ends with Qxf7# or Qf7# around move 4
        let moves = &game_data.moves;
        if let Some(last) = moves.last() {
            let last_upper = last.to_uppercase();
            if (last_upper.contains("QXF7") || last_upper.contains("QF7")
                || last_upper.contains("QXF2") || last_upper.contains("QF2"))
                && last.contains('#')
                && moves.len() <= 8
            {
                self.is_scholars_mate = true;
            }
        }
    }

    fn process_move(&mut self, ctx: &MoveContext) {
        self.total_moves = ctx.move_number;

        if !ctx.is_user_move {
            return;
        }

        let mut pos_after = ctx.board.clone();
        pos_after.play_unchecked(ctx.mv);

        if pos_after.is_checkmate() {
            self.current_mate_move = Some(ctx.move_number);
        }
    }

    fn finish_game(&mut self) -> bool {
        if self.is_scholars_mate {
            return false;
        }

        if let Some(mate_move) = self.current_mate_move {
            let dominated = match self.quickest_moves {
                Some(prev) => mate_move < prev,
                None => true,
            };

            if dominated {
                // Replace previous quickest
                self.quickest_moves = Some(mate_move);
                self.quickest_link = self.current_link.clone();
            }
        }

        false // We only report in matched_game_links
    }

    fn matched_game_links(&self) -> Vec<String> {
        if let Some(ref link) = self.quickest_link {
            vec![link.clone()]
        } else {
            Vec::new()
        }
    }
}
